doc/reverse_engineering/game_mechanics/formations-formulas.md

```diff
@@ -0,0 +1,248 @@
+# Units formations
+
+Military units can be arranged in various formations. Here's an overview on how this may be implemented
+
+### Line formation
+
+If you table the number of units in the first row, in the second row and so on accordingly to the number of units in the group, you'll find a pattern in the units number intervals for a given row count: namely, the length of the first interval `(0...6)` is `6`, the second `(7...15)` is `9 = 6+3 + 0`, the third `(16...28)` is `13 = (6+3) + 3+1`, the fourth (maybe) `(29...46)` is `18 = ((6+3) + 3+1 ) + 3+2`.
+
+The value of `46` is chosen by observing that the difference between the last value for a certain row count and the first value for the following row count is always `3`; in this case, with `46` units `4` rows are formed: the first 3 with 12 units and the last with 10. For the next value, `47`, we guess `5` rows with `10, 10, 10, 10, 7`.
+
+Of course, since there's a limit of 40 to the number of selectable units, there's unfortunately no way to check this guess.
+
+##### Formulas
+
+If we neglect the first interval `(0...6)` from above, we can define a recurrence equation between the others.
+It looks like this:
+
+``` python
+def f(n):
+    if n == 0:
+    	return 9
+    return f(n-1) + n + 3
+```
+
+In order to find the nth last value before the row count changes, we just need to sum f(i) from 0 up to n. Then, in order to get the row count for a specified numOfUnits, we can iteratively check which of the values is the upper bound of the interval where our numOfUnits lies:
+
+``` python
+def findRowCount(numOfUnits):
+    # the last numOfUnits before the row count changes
+    lastBeforeInc = 6  
+    # iteration counter
+    i = 0
+
+    # for the first row
+    if numOfUnits <= 6:
+        return 1
+
+    while numOfUnits > lastBeforeInc: # is numOfUnits greater than last bound?
+        lastBeforeInc += f(i) # no, increase the bound
+        i += 1
+
+    rowCount = i + 1
+    return (lastBeforeInc, rowCount)
+```
+
+Finally, we have to find the number of units for each row. The first row from the list we got from experimentation leads to the following formula:
+
+``` python
+rows[0] = math.ceil(numOfUnits / findRowCount(numOfUnits))
+```
+
+This seems to agree with experimental data and nicely extends to values over 40.
+Rows other from the last one have the same value of the first. The last row is simply:
+
+``` python
+rows[rowCount - 1] = numOfUnits - (rowCount-1)*rows[0]
+```
+
+or, if you care about performance:
+
+``` python
+rows[rowCount - 1] = numOfUnits % (rowCount-1)
+```
+
+**Considerations** : for performance sake (or just out of curiosity), you may want to find a better, non recursive function for `f`. It turns out that the recurrence equation can be solved to
+
+        f(n) = (18 + 7*n + n^2)/2
+
+using Mathematica's
+
+``` mathematica
+        RSolve[{f[n] == f[n-1] + n + 3, f[0] == 9}, f[n], n]
+``` 
```

Comment:
Same problem as above, i.e. you're now recommending more code. This time also in a language that we don't even use in our codebase :D I appreciate what you did here, but it's not really helpful for someone who wants to learn what how to replicate the formation formulas from AoE2. Same goes for the rest of this section which gets too much into details about mathematical specifics that are hard to understand.

I think you should do 2 things to improve this section:

1. Rewrite it so that this section only contains the formulas and their explanation. We don't need much derivation.
2. Create only a small example script which contains the **most readable** mock implmentation for these formulas.